services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/permissions/HankeKayttajaService.kt

```diff
@@ -50,21 +59,37 @@ class HankeKayttajaService(
     }
 
     @Transactional
-    fun saveNewTokensFromApplication(application: ApplicationEntity, hankeId: Int, userId: String) {
+    fun saveNewTokensFromApplication(
+        application: ApplicationEntity,
+        hankeId: Int,
+        hankeTunnus: String,
+        hankeNimi: String,
+        userId: String
+    ) {
         if (featureFlags.isDisabled(Feature.USER_MANAGEMENT)) {
             return
         }
         logger.info {
             "Creating users and user tokens for application ${application.id}, alluid=${application.alluid}}"
         }
-        val contacts =
-            application.applicationData
-                .customersWithContacts()
-                .flatMap { it.contacts }
-                .mapNotNull { userContactOrNull(it.fullName(), it.email) }
+        val applicationIdentifier =
+            application.applicationIdentifier
+                ?: throw ApplicationArgumentException("Application identifier null")
+
+        val inviter = getKayttajaByUserId(hankeId, userId)
+        val contacts = application.applicationData.typedContacts().removeInviter(inviter)
 
         filterNewContacts(hankeId, contacts).forEach { contact ->
-            createTunnisteAndKayttaja(hankeId, contact, userId)
+            createTunnisteAndKayttaja(hankeId, hankeTunnus, hankeNimi, inviter, contact, userId)
+        }
+        contacts.forEach { contact ->
+            sendApplicationNotification(
```

Comment:
Unlike the invitations, these emails are not related to saving new HankeKayttaja. I don't think this should even be in this service.

Reply:
It was placed here as it has a lot of similarities to hanke invitations. But you are making a good point here. HankeKayttajaService should not know anything about application notifications. I'll move that part to application service.

Reply:
Done.